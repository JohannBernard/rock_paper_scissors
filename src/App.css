.App {
  	font-family: 'Poppins', sans-serif;
	font-weight: 400;
	color: #424242;
	min-height: 100vh;
    display: flex;
    flex-direction: column;
}

header{
	display: flex;
	border-bottom: 1px solid #ccc;
	box-shadow: 0 2px 5px 1px rgb(64 60 67 / 16%);
	align-items: center;
}

header h1{
	font-size:28px ;
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
	margin-right: 16px;
}

@media screen and (max-width: 800px) {
	header h1{
		margin: 0;
		font-size:24px ;
	}
}

header .logo_my_coach{
	margin-right: 32px;
}

.headerContent{
	display: flex;
    justify-content: space-between;
    flex: 1;
    margin-right: 16px;
    align-items: center;
	overflow: hidden;
}

.gameSection{
	display: flex; 
	flex-direction: column; 
	justify-content: space-between; 
	flex: 1;
	padding-bottom:32px;
}

.userContainer, .gameContainer{
	display:flex; 
	justify-content:space-around;
	flex: 1;
}

.gameContainer{
	margin-top: 32px;
}

.actions{
	flex-wrap: wrap;
	display:flex; 
	justify-content:space-evenly;
	flex: 1;
}

.random{
	margin-top: 16px;
}

.actionContainer{
	display:block; 
}

.instructionText {
	line-height: 72px;
    font-size: 36px;
	margin:64px 32px 32px;
	font-weight: 600;
	font-style: italic;
	color:#ccc;
	text-align: center;
}

.winLabel{
	font-size: 20px;
	font-style: italic;
    text-align: center;
	margin-bottom: 32px;
}

/* Responsive */
@media screen and (max-width: 800px) {
	header > h1{
		font-size:20px ;
	}
	
	header > .logo_my_coach{
		width: auto;
		height: 60px;
		margin-right: 16px;
	}

	.gameContainer{
		margin-top: 32px;
	}

	.instructionText {
		line-height: 48px;
		font-size: 24px;
		margin:32px 16px 16px;
		font-weight: 600;
		font-style: italic;
		color:#ccc;
		text-align: center;
	}
}

