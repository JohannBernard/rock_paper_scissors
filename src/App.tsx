import { useEffect } from 'react'
import './App.css'
// components
import Button from './components/Button'
import Item from './components/Item'
import Modal from './components/Modal'
import User from './components/User'
// data
import rock_paper_scissors from './data/rock_paper_scissors.json'
import rock_paper_scissors_lizard_spoke from './data/rock_paper_scissors_lizard_spoke.json'
// interface
import IItem from './interfaces/IItem'
// helper
import Utils from './helpers/Utils'
import GameHelper from './helpers/GameHelper'
// hook
import { usePlayer } from './hooks/usePlayer'
import { useComputer } from './hooks/useComputer'
import { useModal } from './hooks/useModal'
import { useGameMode } from './hooks/useGameMode'

function App() {
    // Hook init
    const {
        //Player
        playerScore,
        incrementPlayerScore,
        resetPlayerScore,
        playerItem,
        setPlayerItem,
        resetPlayerItem,
    } = usePlayer()
    const {
        //Computer
        computerScore,
        incrementComputerScore,
        resetComputerScore,
        computerItem,
        setComputerItem,
        resetComputerItem,
    } = useComputer()
    const {
        //Modal
        modalLabel,
        setModalLabel,
        resetModalLabel,
        modalOpen,
        openModal,
        hideModal,
    } = useModal()
    const {
        //Game Mode
        gameMode,
        setGameMode,
    } = useGameMode()

    // Player select an item
    function selectItemAction(item: IItem) {
        const playerItem = item
        const computerItem = gameMode[Utils.getRandomInt(gameMode.length)]
        //Show items
        setPlayerItem(playerItem)
        setComputerItem(computerItem)
        //Increment point to the winner
        if (GameHelper.wonStatus(playerItem, computerItem) === -1) {
            //Player Lost
            incrementComputerScore()
        } else if (GameHelper.wonStatus(playerItem, computerItem) === 1) {
            //Player Win
            incrementPlayerScore()
        }
    }

    // Player select random item
    function randomItemAction() {
        selectItemAction(gameMode[Utils.getRandomInt(gameMode.length)])
    }

    // Toggle game Mode - Add/Remove lizard & spoke
    function toggleGameMode() {
        if (gameMode === rock_paper_scissors) {
            setGameMode(rock_paper_scissors_lizard_spoke)
        } else {
            setGameMode(rock_paper_scissors)
        }
    }

    //Reset Game
    function resetGame() {
        hideModal()
        resetPlayerScore()
        resetComputerScore()
        resetPlayerItem()
        resetComputerItem()
        resetModalLabel()
    }

    // Test if game is finished
    useEffect(() => {
        function testEndGame() {
            if (playerScore === 3 || computerScore === 3) {
                //Fin de la partie
                playerScore === 3
                    ? setModalLabel('Gagn√© !')
                    : setModalLabel('Perdu...')
                openModal()
            }
        }
        testEndGame()
    }, [playerScore, computerScore, openModal, setModalLabel])

    return (
        <div className="App">
            <header>
                <img
                    width={80}
                    height={80}
                    src={require('./assets/images/logo-mycoach.jpeg')}
                    alt="logo_my_coach"
                    className="logo_my_coach"
                />
                <div className="headerContent">
                    <h1>Test MyCoach - Johann Bernard</h1>
                    <Button
                        label={
                            gameMode === rock_paper_scissors
                                ? 'üëä‚úã‚úå'
                                : 'üëä‚úã‚úåü¶éüññ'
                        }
                        onClickCustom={() => toggleGameMode()}
                        icon
                        title="Changer de mode de jeu"
                    />
                </div>
            </header>

            <section className="gameSection">
                <div>
                    <div className="userContainer">
                        <User name="Joueur" score={playerScore} />
                        <User name="Ordinateur" score={computerScore} />
                    </div>

                    <div className="gameContainer">
                        {playerItem || computerItem ? (
                            <>
                                <Item
                                    item={playerItem}
                                    opponent={computerItem}
                                />
                                <Item
                                    item={computerItem}
                                    opponent={playerItem}
                                />
                            </>
                        ) : (
                            <>
                                <p className="instructionText">
                                    Veuillez s√©lectionnez un signe pour
                                    commencer la partie.
                                </p>
                            </>
                        )}
                    </div>
                    <p className="winLabel">
                        {GameHelper.wonLabel(playerItem, computerItem)}
                    </p>
                </div>

                <div className="actionsContainer">
                    <div className="actions">
                        {gameMode.map((item) => (
                            <Button
                                key={item.name}
                                label={item.label}
                                onClickCustom={() => selectItemAction(item)}
                            />
                        ))}
                        <Button
                            label={'Hasard'}
                            onClickCustom={() => randomItemAction()}
                        />
                    </div>
                </div>
            </section>
            <Modal
                label={modalLabel}
                open={modalOpen}
                onClickCustom={() => resetGame()}
            />
        </div>
    )
}

export default App
