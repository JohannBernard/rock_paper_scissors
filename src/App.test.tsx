import React from 'react'
import { render, screen, cleanup } from '@testing-library/react'
import renderer from 'react-test-renderer'
import App from './App'

afterEach(() => {
    cleanup()
})

test('renders texts', () => {
    render(<App />)
    const title = screen.getByText(/Test MyCoach - Johann Bernard/i)
    expect(title).toBeInTheDocument()
    const playerLabel = screen.getByText(/Joueur/i)
    expect(playerLabel).toBeInTheDocument()
    const computerLabel = screen.getByText(/Ordinateur/i)
    expect(computerLabel).toBeInTheDocument()
})

test('matches button snapshot', () => {
    const tree = renderer.create(<App />).toJSON()
    expect(tree).toMatchSnapshot()
})
