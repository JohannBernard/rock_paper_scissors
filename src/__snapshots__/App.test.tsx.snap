// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches button snapshot 1`] = `
<div
  className="App"
>
  <header>
    <img
      alt="logo_my_coach"
      className="logo_my_coach"
      height={80}
      src="logo-mycoach.jpeg"
      width={80}
    />
    <div
      className="headerContent"
    >
      <h1>
        Test MyCoach - Johann Bernard
      </h1>
      <button
        className="my_coach_button my_coach_button_icon"
        data-testid="testButton"
        onClick={[Function]}
        title="Changer de mode de jeu"
      >
        ðŸ‘Šâœ‹âœŒ
      </button>
    </div>
  </header>
  <section
    className="gameSection"
  >
    <div>
      <div
        className="userContainer"
      >
        <div
          className="my_coach_user"
          data-testid=""
        >
          <p
            className="my_coach_user_name"
            data-testid=""
          >
            Joueur
          </p>
          <p
            className="my_coach_user_score"
            data-testid=""
          >
            0
          </p>
        </div>
        <div
          className="my_coach_user"
          data-testid=""
        >
          <p
            className="my_coach_user_name"
            data-testid=""
          >
            Ordinateur
          </p>
          <p
            className="my_coach_user_score"
            data-testid=""
          >
            0
          </p>
        </div>
      </div>
      <div
        className="gameContainer"
      >
        <p
          className="instructionText"
        >
          Veuillez sÃ©lectionnez un signe pour commencer la partie.
        </p>
      </div>
      <p
        className="winLabel"
      />
    </div>
    <div
      className="actionsContainer"
    >
      <div
        className="actions"
      >
        <button
          className="my_coach_button"
          data-testid="testButton"
          onClick={[Function]}
        >
          Pierre
        </button>
        <button
          className="my_coach_button"
          data-testid="testButton"
          onClick={[Function]}
        >
          Feuille
        </button>
        <button
          className="my_coach_button"
          data-testid="testButton"
          onClick={[Function]}
        >
          Ciseaux
        </button>
        <button
          className="my_coach_button"
          data-testid="testButton"
          onClick={[Function]}
        >
          Hasard
        </button>
      </div>
    </div>
  </section>
  <div
    className="modal_window"
    data-testid=""
    id="open-modal"
  >
    <div
      className="modal_container"
    >
      <div
        className="modal_label"
        data-testid=""
      />
      <button
        className="my_coach_button"
        data-testid="testButton"
        onClick={[Function]}
      >
        Recommencer
      </button>
    </div>
  </div>
</div>
`;
